File: mavsim_python/message_types/msg_waypoints.py
"""
msg_waypoints
    - messages type for input to path manager
    
part of mavsim_python
    - Beard & McLain, PUP, 2012
    - Last update:
        3/26/2019 - RWB
        3/31/2020 - RWB
"""
import numpy as np


class MsgWaypoints:
    def __init__(self):
        # the first two flags are used for interacting with the path planner
        #
        # flag to indicate waypoints recently changed (set by planner)
        self.flag_waypoints_changed = True
        self.plot_updated = False  # used to plot waypoints

        # type of waypoint following:
        #   - straight line following
        #   - fillets between straight lines
        #   - follow dubins paths
        self.type = 'straight_line'
        # self.type = 'fillet'
        # self.type = 'dubins'
        # current number of valid waypoints in memory
        self.num_waypoints = 0
        # [n, e, d] - coordinates of waypoints
        self.ned = np.array([[],[],[]])
        # the airspeed that is commanded along the waypoints
        self.airspeed = np.array([])
        # the desired course at each waypoint (used only for Dubins paths)
        self.course = np.array([])

        # these last three variables are used by the path planner running cost at each node
        self.cost = np.array([])
        # index of the parent to the node
        self.parent = np.array([])
        # can this node connect to the goal?
        self.connect_to_goal = np.array([])

    def add(self, ned=np.array([[0, 0, 0]]).T, airspeed=0,
            course=np.inf, cost=0, parent=0, connect_to_goal=0):
        self.num_waypoints = self.num_waypoints + 1
        self.ned = np.append(self.ned, ned, axis=1)
        self.airspeed = np.append(self.airspeed, airspeed)
        self.course = np.append(self.course, course)
        self.cost = np.append(self.cost, cost)
        self.parent = np.append(self.parent, parent)
        self.connect_to_goal = np.append(self.connect_to_goal, connect_to_goal)

File: mavsim_matlab/message_types/msg_waypoints.m
% msg_waypoints
%   - message type for path planning
%
% mavsim_matlab 
%     - Beard & McLain, PUP, 2012
%     - Update history:  
%         4/8/2019 - RWB
classdef msg_waypoints < handle
   %--------------------------------
    properties
        flag_waypoints_changed
        flag_manager_requests_waypoints
        type
        max_waypoints
        num_waypoints
        ned
        airspeed
        course
        cost
        parent
        flag_connect_to_goal
    end
    %--------------------------------
    methods
        %------constructor-----------
        function self = msg_waypoints()
            % the first two flags are used for interacting with the path
            % planner
            %
            % flag to indicate waypoints recently changed (set by planner)
            self.flag_waypoints_changed = 1;  
            
            % type of waypoint following:
            %   - straight line following
            %   - fillets between straight lines
            %   - follow dubins paths
            self.type = 'straight_line';
            %self.type = 'fillet';
            %self.type = 'dubins';
            % current number of valid waypoints in memory
            self.num_waypoints = 0;
            % [n, e, d] - coordinates of waypoints
            self.ned = [];
            % the airspeed that is commanded along the waypoints
            self.airspeed = [];
            % the desired course at each waypoint (used only for Dubins
            % paths)
            self.course = [];
            
            % these last three variables are used by the path planner
            % running cost at each node
            self.cost = [];
            % index of the parent to the node
            self.parent = [];   
            % can this node connect to the goal?  1==connected, 0==not
            % connected
            self.flag_connect_to_goal = [];  
        end
        %---------------------------------------------------
        function self = add(self, ned, airspeed, course,...
                            cost, parent, connect_to_goal)
            self.num_waypoints = self.num_waypoints + 1;
            self.ned = [self.ned, ned];  % ned is 3x1 position vector
            self.airspeed = [self.airspeed, airspeed];
            self.course = [self.course, course];
            self.cost = [self.cost, cost];
            self.parent = [self.parent, parent];
            self.flag_connect_to_goal = [self.flag_connect_to_goal, connect_to_goal];
        end
    end
end